src/org/zaproxy/zap/authentication/PostBasedAuthenticationMethodType.java

```diff
@@ -252,6 +260,43 @@ protected HttpMessage prepareRequestMessage(UsernamePasswordAuthenticationCreden
 			return requestMessage;
 		}
 
+		private void replaceAntiCsrfTokenAndCookieValueIfRequired(HttpMessage requestMessage) throws URIException, NullPointerException, HttpMalformedHeaderException {
+			if(extAntiCsrf == null) {
+				extAntiCsrf = Control.getSingleton().getExtensionLoader().getExtension(ExtensionAntiCSRF.class);
+			}
+			List<AntiCsrfToken> antiCsrfTokensOfFirstLoginMsg = null;
+			if(extAntiCsrf != null) {
+				antiCsrfTokensOfFirstLoginMsg = extAntiCsrf.getTokens(requestMessage);
+			}
+			AntiCsrfToken antiCsrfTokenOfFirstLoginMsg = null;
+			if(antiCsrfTokensOfFirstLoginMsg != null && antiCsrfTokensOfFirstLoginMsg.size() > 0) {
+				antiCsrfTokenOfFirstLoginMsg = antiCsrfTokensOfFirstLoginMsg.get(0);
+			}
+			
+			if (antiCsrfTokenOfFirstLoginMsg != null) {
+				this.requireAnitCsrfTokenWhenLogin = true;
+				URI loginURI = new URI(loginRequestURL, true, "UTF-8");
+				HttpMessage loginMsgToRegenerateAntiCsrfToken = new HttpMessage(loginURI);
+				loginMsgToRegenerateAntiCsrfToken.setRequestingUser(null);
```

Comment:
Why not the user? We do want to use the same session.

Reply:
Initially, I put this to prevent cookies of `POST` request to `login.php`getting replaced by old values instead of the new values from the response of `GET` request to `login.php`.  When refactoring I messed up.
Is there any way to get the user from this scope?
As the first `GET`, `POST` to `login.php` is made from the browser when `forcedUser` mode is disabled, they do not have the `user` info set.

Reply:
We need to pass the user from `authenticate` to `prepareRequestMessage` and then to this method (we could also call `replaceAntiCsrfTokenAndCookieValueIfRequired` from `authenticate` after preparing the message).

Yeah, that's not a problem, we just need that when re-authenticating.